src/ast/ast_validations.c

```diff
@@ -906,55 +906,72 @@ static AST_Validation _Validate_MERGE_Clauses(const AST *ast) {
-		 * MATCH (a) CREATE (a)
-		 * But this is a no-op query, and we don't have the logic to differentiate this from a valid query like
-		 * MATCH (a) CREATE (a)-[:E]->(:B) */
-		for(uint j = 1; j < nelems; j += 2) {
-			const cypher_astnode_t *rel = cypher_ast_pattern_path_get_element(path, j);
-			const cypher_astnode_t *identifier = cypher_ast_rel_pattern_get_identifier(rel);
-			// Validate that no relation aliases are previously bound.
-			if(identifier) {
-				const char *alias = cypher_ast_identifier_get_name(identifier);
-				printf("\nDA %s\n", alias);
-				if(raxFind(defined_aliases, (unsigned char *)alias, strlen(alias)) != raxNotFound) {
-					ErrorCtx_SetError("The bound variable %s' can't be redeclared in a CREATE clause", alias);
-					return AST_INVALID;
-				}
-			}
-
-			// Validate that each relation has exactly one type.
-			uint reltype_count = cypher_ast_rel_pattern_nreltypes(rel);
-			if(reltype_count != 1) {
-				ErrorCtx_SetError("Exactly one relationship type must be specified for CREATE");
+	if(_ValidateInlinedPropertiesOnPath(path) != AST_VALID) return AST_INVALID;
+	uint nelems = cypher_ast_pattern_path_nelements(path);
+	/* Visit every relationship (every odd offset) on the path to validate its alias and structure.
+	 * TODO There should also be a syntax error for redeclaring nodes, as in:
+	 * MATCH (a) CREATE (a)
+	 * But this is a no-op query, and we don't have the logic to differentiate this from a valid query like
+	 * MATCH (a) CREATE (a)-[:E]->(:B) */
+	for(uint j = 1; j < nelems; j += 2) {
+		const cypher_astnode_t *rel = cypher_ast_pattern_path_get_element(path, j);
+		const cypher_astnode_t *identifier = cypher_ast_rel_pattern_get_identifier(rel);
+		// Validate that no relation aliases are previously bound.
+		if(identifier) {
+			const char *alias = cypher_ast_identifier_get_name(identifier);
+			if(raxFind(defined_aliases, (unsigned char *)alias, strlen(alias)) != raxNotFound) {
+				ErrorCtx_SetError("The bound variable '%s' can't be redeclared in a CREATE clause", alias);
 				return AST_INVALID;
 			}
+		}
 
-			// Validate that each relation being created is directed.
-			if(cypher_ast_rel_pattern_get_direction(rel) == CYPHER_REL_BIDIRECTIONAL) {
-				ErrorCtx_SetError("Only directed relationships are supported in CREATE");
-				return AST_INVALID;
-			}
+		// Validate that each relation has exactly one type.
+		uint reltype_count = cypher_ast_rel_pattern_nreltypes(rel);
+		if(reltype_count != 1) {
+			ErrorCtx_SetError("Exactly one relationship type must be specified for CREATE");
+			return AST_INVALID;
+		}
+
+		// Validate that each relation being created is directed.
+		if(cypher_ast_rel_pattern_get_direction(rel) == CYPHER_REL_BIDIRECTIONAL) {
+			ErrorCtx_SetError("Only directed relationships are supported in CREATE");
+			return AST_INVALID;
 		}
 	}
 
 	return AST_VALID;
 }
 
+static void _AST_Path_GetDefinedIdentifiers(const cypher_astnode_t *path, rax *identifiers) {
```

Comment:
no need to move this function

Reply:
I added a forward declaration so we wont have to move it